libraries/eosiolib/contracts/eosio/key_value.hpp

```diff
@@ -417,36 +463,68 @@ class kv_table {
       }
 
       bool operator==(const iterator& b) const {
-         return compare(b) == 0;
+         return base_iterator::compare(b) == 0;
       }
 
       bool operator!=(const iterator& b) const {
-         return compare(b) != 0;
+         return base_iterator::compare(b) != 0;
       }
 
       bool operator<(const iterator& b) const {
-         return compare(b) < 0;
+         return base_iterator::compare(b) < 0;
       }
 
       bool operator<=(const iterator& b) const {
-         return compare(b) <= 0;
+         return base_iterator::compare(b) <= 0;
       }
 
       bool operator>(const iterator& b) const {
-         return compare(b) > 0;
+         return base_iterator::compare(b) > 0;
       }
 
       bool operator>=(const iterator& b) const {
-         return compare(b) >= 0;
+         return base_iterator::compare(b) >= 0;
       }
+   };
 
-   private:
-      uint32_t itr;
-      status itr_stat;
+   class reverse_iterator : public base_iterator {
+      using base_iterator::itr;
+      using base_iterator::itr_stat;
+      using base_iterator::index;
 
-      const kv_index* index;
+   public:
+      using status = typename base_iterator::status;
+
+      using base_iterator::base_iterator;
+
+      reverse_iterator(iterator&& other) : base_iterator{std::move(other)} {}
+
+      reverse_iterator& operator++() {
+         eosio::check(itr_stat != status::iterator_end, "incremented past the end");
+         itr_stat = static_cast<status>(internal_use_do_not_use::kv_it_prev(itr));
+         return *this;
+      }
 
-      int compare(const iterator& b) const {
+      reverse_iterator& operator--() {
+         if (!itr) {
+            itr = internal_use_do_not_use::kv_it_create(index->tbl->db_name, index->contract_name.value, index->prefix.data(), index->prefix.size());
+            itr_stat = static_cast<status>(internal_use_do_not_use::kv_it_lower_bound(itr, "", 0));
```

Comment:
This falls through and also calls next, which will leave the iterator in the wrong position.